source/h2_stream.c

```diff
@@ -261,22 +287,44 @@ static void s_stream_cross_thread_work_task(struct aws_channel_task *task, void
         /* window_update_size is ensured to be not greater than AWS_H2_WINDOW_UPDATE_MAX */
         window_update_size = stream->synced_data.window_update_size;
         stream->synced_data.window_update_size = 0;
+        reset_called = stream->synced_data.reset_called;
+        user_reset_error_code = stream->synced_data.user_reset_error_code;
 
         s_unlock_synced_data(stream);
     } /* END CRITICAL SECTION */
 
     while (!aws_linked_list_empty(&pending_frames)) {
         struct aws_linked_list_node *node = aws_linked_list_pop_front(&pending_frames);
         struct aws_h2_frame *frame = AWS_CONTAINER_OF(node, struct aws_h2_frame, node);
+        if (ignore_window_udpate && (frame->type == AWS_H2_FRAME_T_WINDOW_UPDATE)) {
+            /* Do not send the window udpate frame and clean it up */
+            aws_h2_frame_destroy(frame);
+            continue;
+        }
         aws_h2_connection_enqueue_outgoing_frame(connection, frame);
     }
     /* The largest legal value will be 2 * max window size, which is way less than INT64_MAX, so if the window_size_self
      * overflows, remote peer will find it out. So just apply the change and ignore the possible overflow.*/
     stream->thread_data.window_size_self += window_update_size;
 
+    if (reset_called) {
+        /* TODO:: It's possible that server sends a rst_stream with NO_ERROR, after full response has been sent. In this
+         * case the aws_code here should be AWS_ERROR_HTTP_SUCCESS */
+        struct aws_h2err h2err;
+        h2err.h2_code = user_reset_error_code;
+        h2err.aws_code = AWS_ERROR_HTTP_RST_STREAM_SENT;
```

Comment:
we could totally handle this TODO righ now
```
if (is_server && stream->thread_data.state == HALF_CLOSED_LOCAL)
    /* (RFC-7540 8.1) A server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response */
    h2err.aws_code = AWS_ERROR_SUCCESS;
else:
    h2err.aws_code = AWS_ERROR_HTTP_RST_STREAM_SENT
```

Reply:
I don't know... I just feel that this kind of check will be more accurate if happens when the user calls the reset_stream function, instead of when the task runs.

Reply:
Changed my mind, the edge case is just too crazy to happen in real world. Will do.